hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-2.x-replication/src/main/java/com/google/cloud/bigtable/hbase2_x/replication/HbaseToCloudBigtableReplicationEndpoint.java

```diff
@@ -17,40 +17,119 @@
 import org.apache.hadoop.hbase.wal.WAL;
+import org.apache.hbase.thirdparty.com.google.common.collect.Lists;
+import org.apache.hbase.thirdparty.com.google.common.util.concurrent.AbstractService;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-// TODO(remove BaseReplicationEndpoint extension).
 @InternalExtensionOnly
-public class HbaseToCloudBigtableReplicationEndpoint extends BaseReplicationEndpoint {
+public class HbaseToCloudBigtableReplicationEndpoint extends AbstractService
+    implements ReplicationEndpoint {
 
   private static final Logger LOG =
       LoggerFactory.getLogger(HbaseToCloudBigtableReplicationEndpoint.class);
 
   private final CloudBigtableReplicator cloudBigtableReplicator;
   private final HBaseMetricsExporter metricsExporter;
 
+  protected Context ctx;
+
   public HbaseToCloudBigtableReplicationEndpoint() {
     cloudBigtableReplicator = new CloudBigtableReplicator();
     metricsExporter = new HBaseMetricsExporter();
   }
 
+  @Override
+  public void init(Context context) throws IOException {
+    this.ctx = context;
+    if (this.ctx != null) {
+      ReplicationPeer peer = this.ctx.getReplicationPeer();
+      if (peer != null) {
+        peer.registerPeerConfigListener(this);
+      } else {
+        LOG.warn(
+            "Not tracking replication peer config changes for Peer Id "
+                + this.ctx.getPeerId()
+                + " because there's no such peer");
+      }
+    }
+  }
+
+  @Override
+  public boolean canReplicateToSameCluster() {
+    return false;
+  }
+
   @Override
   public UUID getPeerUUID() {
     return cloudBigtableReplicator.getPeerUUID();
   }
 
+  @Override
+  public WALEntryFilter getWALEntryfilter() {
+    ArrayList<WALEntryFilter> filters = Lists.newArrayList();
+    WALEntryFilter scopeFilter = getScopeWALEntryFilter();
+    if (scopeFilter != null) {
+      filters.add(scopeFilter);
+    }
+    WALEntryFilter tableCfFilter = getNamespaceTableCfWALEntryFilter();
+    if (tableCfFilter != null) {
+      filters.add(tableCfFilter);
+    }
+    if (ctx != null && ctx.getPeerConfig() != null) {
+      String filterNameCSV =
+          ctx.getPeerConfig().getConfiguration().get(REPLICATION_WALENTRYFILTER_CONFIG_KEY);
+      if (filterNameCSV != null && !filterNameCSV.isEmpty()) {
+        String[] filterNames = filterNameCSV.split(",");
+        for (String filterName : filterNames) {
+          try {
+            Class<?> clazz = Class.forName(filterName);
+            filters.add((WALEntryFilter) clazz.getDeclaredConstructor().newInstance());
+          } catch (Exception e) {
+            LOG.error("Unable to create WALEntryFilter " + filterName, e);
+          }
+        }
+      }
+    }
+    return filters.isEmpty() ? null : new ChainWALEntryFilter(filters);
+  }
+
+  /**
+   * Returns a WALEntryFilter for checking the scope. Subclasses can return null if they don't want
+   * this filter
+   */
+  protected WALEntryFilter getScopeWALEntryFilter() {
```

Comment:
What is a scope filter? Why do we want it?

Reply:
REPLICATION_SCOPE_GLOBAL vs REPLICATION_SCOPE_LOCAL.

it excludes the key values which are scoped to local. https://github.com/apache/hbase/blob/master/hbase-common/src/main/java/org/apache/hadoop/hbase/HConstants.java#L741